src/client/executor.rs

```diff
@@ -103,16 +118,47 @@ impl Client {
             Err(err) => {
                 self.inner
                     .topology
-                    .handle_post_handshake_error(err.clone(), conn, server)
+                    .handle_post_handshake_error(err.clone(), &conn, server)
                     .await;
-                // TODO RUST-90: Do not retry if session is in a transaction
-                if self.inner.options.retry_reads == Some(false)
-                    || !op.is_read_retryable()
-                    || !err.is_read_retryable()
+
+                if let ErrorKind::CommandError(ref err) = err.kind.as_ref() {
+                    if err.code == 20 && err.message.starts_with("Transaction numbers") {
+                        let mut err = err.clone();
+                        err.message = "This MongoDB deployment does not support retryable writes. \
+                                       Please add retryWrites=false to your connection string."
+                            .to_string();
+                        return Err(ErrorKind::CommandError(err).into());
+                    }
+                }
+
+                // For a pre-4.4 connection, an error label should be added to any write-retryable
+                // error as long as the retry_writes client option is not set to false. For a 4.4+
+                // connection, a label should be added only to non-command errors.
+                let err = if self.inner.options.retry_writes != Some(false)
+                    && op.retryability() == Retryability::Write
                 {
-                    return Err(err);
+                    let add_to_command_error = conn
+                        .stream_description()?
+                        .max_wire_version
+                        .map_or(false, |version| version <= 8);
+                    err.with_retryable_write_label(add_to_command_error)
```

Comment:
Since we'll be cloning anyways, we can use the "clone on write" behavior of [`make_mut`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.make_mut).

Reply:
It's probably more straightforward to not rely on that actually, so we can just leave this as-is.